### PR TITLE
WIP SASS-6114: Scaffold Repairs and Maintenance

### PR DESCRIPTION
**Pull request by lukastymo**

### CODE REVIEW

File: test/controllers/journeys/expenses/repairsandmaintenance/RepairsAndMaintenanceAmountControllerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import play.api.mvc.Request
+import play.twirl.api.{Html, HtmlFormat}
+import views.html.journeys.expenses.repairsandmaintenance.RepairsAndMaintenanceAmountView
+
+class RepairsAndMaintenanceAmountControllerSpec
+    extends BigDecimalGetAndPostQuestionBaseSpec("RepairsAndMaintenanceAmountController", RepairsAndMaintenanceAmountPage) {
+  lazy val onPageLoadRoute = RepairsAndMaintenanceAmountController.onPageLoad(taxYear, stubbedBusinessId, NormalMode).url
+  lazy val onSubmitRoute   = RepairsAndMaintenanceAmountController.onSubmit(taxYear, stubbedBusinessId, NormalMode).url
+
+  def createForm(userType: UserType): Form[BigDecimal] = new RepairsAndMaintenanceAmountFormProvider()(userType)
```
Comment (harry-ritchie) [inline test/controllers/journeys/expenses/repairsandmaintenance/RepairsAndMaintenanceAmountControllerSpec.scala:38]: Could we add `override` to these methods to be clear they are overidden?

File: test/base/ControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    when(mockSessionRepository.set(any())) thenReturn Future.successful(true)
+
+  }
+
+  def onwardRoute = Call("GET", "/foo")
+
+  private def createApp(userType: UserType, answers: Option[UserAnswers], mockSessionRepository: SessionRepository) = {
+    applicationBuilder(answers, isAgent(userType.toString))
+      .overrides(
+        bind[ExpensesNavigator].toInstance(new FakeExpensesNavigator(onwardRoute)),
```
Comment (harry-ritchie) [inline test/base/ControllerSpec.scala:64]: Is there some way we could make this not specific to Expenses so this trait can be used for all journeys?

File: test/base/ControllerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    (Language.Welsh, UserType.Agent)
+  )
+
+  abstract class TestApp(userType: UserType, answers: Option[UserAnswers]) {
+    private val mockSessionRepository = mock[SessionRepository]
+
+    implicit val application              = createApp(userType, answers, mockSessionRepository)
+    implicit val messagesApi: MessagesApi = application.injector.instanceOf[MessagesApi]
+
+    private val onwardRoute = Call("GET", "/foo")
```
Comment (harry-ritchie) [inline test/base/ControllerSpec.scala:53]: `onwardRoute` is not used anywhere.

File: test/base/ControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import play.api.inject.bind
+
+import scala.concurrent.Future
+import scala.reflect.ClassTag
+
+trait ControllerSpec extends SpecBase with MockitoSugar with TableDrivenPropertyChecks {
+
+  val langUserTypeCases = Table(
+    ("Language", "userType"),
+    (Language.English, UserType.Individual: UserType),
```
Comment (harry-ritchie) [inline test/base/ControllerSpec.scala:41]: Don't need the `:UserType here`

File: test/base/BigDecimalGetAndPostQuestionBaseSpec.scala [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            status(result) mustEqual OK
+            contentAsString(result) mustEqual renderView(getRequest, messages(application), application)(
+              form,
+              NormalMode,
+              taxYear,
+              stubbedBusinessId).toString
+          }
+        }
+
+        "Return OK for a GET if previous answer exists" in new TestApp(userType, Some(filledUserAnswers)) {
```
Comment (harry-ritchie) [inline test/base/BigDecimalGetAndPostQuestionBaseSpec.scala:71]: Maybe rephrase to `Return OK for a GET if a previous answer exists, with the view populated with the previous answer`?

File: test/base/BigDecimalGetAndPostQuestionBaseSpec.scala [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+  def getRequest  = FakeRequest(GET, onPageLoadRoute)
+  def postRequest = FakeRequest(POST, onSubmitRoute).withFormUrlEncodedBody(("value", validAnswer.toString))
+
+  forAll(langUserTypeCases) { case (lang, userType) =>
+    s"$controllerName for ${lang.entryName} and ${userType}" - {
+      val form: Form[BigDecimal] = createForm(userType)
+
+      "Loading page" - {
+        "Redirect to Journey Recover for a GET if existing data not found " in new TestApp(userType, None) {
```
Comment (harry-ritchie) [inline test/base/BigDecimalGetAndPostQuestionBaseSpec.scala:51]: Maybe rephrase as `Redirect to Journey Recover for a GET if **prerequisite** data not found`, as the wording to me is very similar to the test below.

File: test/base/BigDecimalGetAndPostQuestionBaseSpec.scala [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+              NormalMode,
+              taxYear,
+              stubbedBusinessId).toString
+          }
+
+        }
+      }
+
+      "Submitting page" - {
+        "Redirect to Journey Recovery for POST if no existing data is found" in new TestApp(userType, None) {
```
Comment (harry-ritchie) [inline test/base/BigDecimalGetAndPostQuestionBaseSpec.scala:86]: `..no prerequisite data is found`?

File: test/base/BigDecimalGetAndPostQuestionBaseSpec.scala [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+import play.api.test.Helpers._
+import play.twirl.api.HtmlFormat
+
+abstract case class BigDecimalGetAndPostQuestionBaseSpec(controllerName: String, page: QuestionPage[BigDecimal]) extends ControllerSpec {
+  val onPageLoadRoute: String
+  val onSubmitRoute: String
+
+  def createForm(userType: UserType): Form[BigDecimal]
+
+  def renderView(implicit request: Request[_], messages: Messages, application: Application): (Form[_], Mode, Int, String) => HtmlFormat.Appendable
```
Comment (harry-ritchie) [inline test/base/BigDecimalGetAndPostQuestionBaseSpec.scala:38]: If we were to use this method for other BigDecimal question pages, the inputs to the views may slightly change, e.g. one may require accountingType passed into the view, and another may not (this is also the case with the form providers). What would be the best way to accommodate for this in the base class?

File: test/controllers/journeys/expenses/repairsandmaintenance/RepairsAndMaintenanceAmountControllerSpec.scala (hunk 1 above)
Comment (lukastymo) [reply test/controllers/journeys/expenses/repairsandmaintenance/RepairsAndMaintenanceAmountControllerSpec.scala:38]: the rule I used before about override in Scala was:
- do not use it if you add implementation (implement trait)
- add override if there is some implementation already and you really change, override behavior.

Here we don't change any implementation so override is omitted.

Let me know if you like it.

File: test/base/ControllerSpec.scala (hunk 3 above)
Comment (lukastymo) [reply test/base/ControllerSpec.scala:53]: used in BigDecimalGetAndPostQuestionBaseSpec - I think we mock navigation and test if we redirect to this route - it may be common for all ControllerSpec

File: test/base/ControllerSpec.scala (hunk 4 above)
Comment (lukastymo) [reply test/base/ControllerSpec.scala:41]: done

File: test/base/BigDecimalGetAndPostQuestionBaseSpec.scala (hunk 5 above)
Comment (lukastymo) [reply test/base/BigDecimalGetAndPostQuestionBaseSpec.scala:71]: it's a bit long and too descriptive, but I can do it to be precise what the test does

File: test/base/BigDecimalGetAndPostQuestionBaseSpec.scala (hunk 6 above)
Comment (lukastymo) [reply test/base/BigDecimalGetAndPostQuestionBaseSpec.scala:51]: done

File: test/base/BigDecimalGetAndPostQuestionBaseSpec.scala (hunk 7 above)
Comment (lukastymo) [reply test/base/BigDecimalGetAndPostQuestionBaseSpec.scala:86]: done

File: test/controllers/journeys/expenses/repairsandmaintenance/RepairsAndMaintenanceAmountControllerSpec.scala (hunk 1 above)
Comment (harry-ritchie) [reply test/controllers/journeys/expenses/repairsandmaintenance/RepairsAndMaintenanceAmountControllerSpec.scala:38]: Makes sense!

File: test/base/ControllerSpec.scala (hunk 2 above)
Comment (lukastymo) [reply test/base/ControllerSpec.scala]: will do

File: test/base/BigDecimalGetAndPostQuestionBaseSpec.scala (hunk 8 above)
Comment (lukastymo) [reply test/base/BigDecimalGetAndPostQuestionBaseSpec.scala:41]: ok, made a fix to accommodate accountingType and other more generic Types. We can pass now TestScenario

**Review by harry-ritchie — APPROVED** [review]
